# Descriptive Analysis – World Happiness Health Analytics

## Notebook Overview
This descriptive analysis examines the ** merged World Happiness dataset from 2015–2024 ** with **1,320 rows** of data across **132 countries**. The analysis focuses on understanding patterns, distributions, and relationships in global happiness and health metrics.


**Dataset Columns**:  
- Country  
- Happiness Rank  
- Happiness Score  
- Economy  
- Family  
- Healthy Life Expectancy  
- Freedom to Make Life Choices  
- Perceptions of Corruption  
- Generosity  
- Continent  

---

## Analysis Framework: Four Key Descriptive Charts

### 1. Health vs Happiness Relationship Analysis
- **Type of Analysis**: Correlation Analysis (Bivariate-Statistical technique that examines the relationship between two continuous variables.)  
- **Chart Type**: Scatter plot with trend line and continent color-coding  
- **Purpose**: Explores the direct relationship between healthy life expectancy and happiness scores across countries  

**Variables**:
- X-axis: Healthy Life Expectancy  
- Y-axis: Happiness Score  
- Color: Continent  

**Expected Insights**:
- Strength of health-happiness correlation  
- Regional patterns in health-happiness relationship  
- Identification of outlier countries  
- Linear vs non-linear relationship patterns  

---

### 2. Happiness Score Trends Over Time (2015–2024)
- **Type of Analysis**: Time Series Analysis (Temporal Trend)  
- **Chart Type**: Multi-line chart or area chart  
- **Purpose**: Reveals temporal patterns and global happiness trajectory over the decade  

**Variables**:
- X-axis: Year (2015–2024)  
- Y-axis: Happiness Score  
- Multiple lines: Top countries or continental averages  

**Expected Insights**:
- Global happiness trend direction (improving/declining)  
- Year-over-year changes and volatility  
- Identification of significant events affecting happiness  
- Comparative country/regional performance over time  

---

### 3. Key Happiness Factors Correlation Matrix
- **Type of Analysis**: Correlation Matrix Analysis (Multivariate-Comprehensive analysis examining relationships between multiple variables simultaneously.)  
- **Chart Type**: Correlation heatmap  
- **Purpose**: Identifies which factors most strongly influence happiness and their inter-relationships  

**Variables**:
- Happiness Score  
- Economy  
- Family  
- Healthy Life Expectancy  
- Freedom to Make Life Choices  
- Perceptions of Corruption  
- Generosity  

**Expected Insights**:
- Strongest predictors of happiness  
- Factor interdependencies  
- Potential multicollinearity issues  
- Comprehensive relationship mapping  

---

### 4. Continental Health & Happiness Comparison
- **Type of Analysis**: Comparative Analysis (Cross-sectional)  
- **Chart Type**: Grouped bar chart or dual-axis chart  
- **Purpose**: Compares both health and happiness metrics across continents  

**Variables**:
- X-axis: Continent  
- Y-axis 1: Average Happiness Score  
- Y-axis 2: Average Healthy Life Expectancy  
- Representation: Grouped bars or dual axes  

**Expected Insights**:
- Regional performance rankings  
- Health-happiness gaps by continent  
- Identification of best/worst performing regions  
- Regional disparities and patterns  

---

### 📊 Chart Outputs
- **Scatter Plot**: Health vs Happiness with continental grouping  
- **Time Series Plot**: Happiness trends over 2015–2024  
- **Correlation Heatmap**: All factor relationships  
- **Comparative Bar Chart**: Continental performance comparison  


---

## Next Steps
This descriptive analysis will provide the foundation for:
- **Diagnostic Analysis**: Investigating why certain patterns exist  
- **Predictive Analysis**: Forecasting future happiness trends  
- **Prescriptive Analysis**: Recommending policy interventions  
- **Machine Learning Models**: Building predictive algorithms  

---

## Tools and Libraries
- **Python**: Primary programming language  
- **Pandas**: Data manipulation and analysis  
- **Matplotlib / Seaborn**: Statistical visualizations  
- **Plotly**: Interactive charts  
- **NumPy**: Numerical computations  
- **Scipy**: Statistical analysis functions  


In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
from scipy import stats  # For correlation Analysis

In [2]:
df =pd.read_csv('../data/cleaned/cleaned_df_all_2015_to_2024.csv')

In [ ]:
# Initial data exploration
df.info()
df.describe()
df.isnull().sum()
df.head()


### Fixing Congo Variants (Row Mismatch 1318 → 1310)

### 🎯 Goal
- **Expected rows:** 1,310 ( 31 countries × 10 years, 2015‑2024 )  
- **Actual rows:** 1,318  
- **Cause:** *Congo* appears **18 times** instead of the expected 10.  
  - Two spellings exist:
    1. **Congo (Brazzaville)**  → mapped to `Congo`
    2. **Congo (Kinshasa)**  → mapped to `Congo`
- One of these variants is **missing in ≥ 2 years**, so its extra rows must be removed (8 rows) to restore the 1,310‑row target.

---

### 🗺️ Plan of Attack

1. **Locate both original variants** before mapping.  
2. **Count each variant** by year (2015‑2024).  
3. **Identify the variant that does *not* appear in all 10 years.**  
4. **Drop all rows** of that incomplete variant (≈ 8 rows).  
5. **Verify final row count** = 1,310.

In [39]:
# Total number of unique countries
unique_countries_count = df['Country'].nunique()
print(f"Total unique countries: {unique_countries_count}")

# Number of times each country appears
country_counts = df['Country'].value_counts()

print(country_counts.to_string())



Total unique countries: 131
Country
Congo                      18
Afghanistan                10
Netherlands                10
Poland                     10
Philippines                10
Peru                       10
Paraguay                   10
Panama                     10
Palestinian Territories    10
Pakistan                   10
Norway                     10
North Macedonia            10
Nigeria                    10
Niger                      10
Nicaragua                  10
New Zealand                10
Nepal                      10
Romania                    10
Myanmar                    10
Morocco                    10
Montenegro                 10
Mongolia                   10
Moldova                    10
Mexico                     10
Mauritius                  10
Mauritania                 10
Malta                      10
Mali                       10
Malaysia                   10
Malawi                     10
Madagascar                 10
Luxembourg                 10
Port

In [40]:
# Find duplicate rows based on Country and Year for Congo
duplicate_rows = df[df['Country'] == 'Congo'].duplicated(subset=['Country', 'Year'], keep=False)

# Show duplicated rows only
print(df[(df['Country'] == 'Congo') & (duplicate_rows)])


    Country  Happiness Rank  Happiness Score   Economy    Family  \
240   Congo             120            4.517  0.000000  1.001200   
241   Congo             139            3.989  0.678660  0.662900   
242   Congo             125            4.272  0.056610  0.806760   
243   Congo             127            4.236  0.771090  0.477990   
244   Congo             124            4.291  0.808964  0.832044   
245   Congo             126            4.280  0.092102  1.229023   
246   Congo             114            4.559  0.682000  0.811000   
247   Congo             132            4.245  0.069000  1.136000   
248   Congo             103            4.812  0.673000  0.799000   
249   Congo             127            4.418  0.094000  1.125000   
250   Congo              88            5.190  0.630000  0.760000   
251   Congo             131            4.310  0.060000  0.830000   
254   Congo              86            5.267  0.921000  0.665000   
255   Congo             133            3.207  0.

In [41]:
# Load each CSV file into a separate DataFrame
df_2015 = pd.read_csv('../data/raw/2015.csv')  # Load the 2015 World Happiness data
df_2016 = pd.read_csv('../data/raw/2016.csv')  # Load the 2016 World Happiness data
df_2017 = pd.read_csv('../data/raw/2017.csv')  # Load the 2017 World Happiness data
df_2018 = pd.read_csv('../data/raw/2018.csv')  # Load the 2018 World Happiness data
df_2019 = pd.read_csv('../data/raw/2019.csv')  # Load the 2019 World Happiness data
df_2020 = pd.read_csv('../data/raw/2020.csv')  # Load the 2020 World Happiness data
df_2021 = pd.read_csv('../data/raw/2021.csv')  # Load the 2021 World Happiness data
df_2022 = pd.read_csv('../data/raw/2022.csv')  # Load the 2022 World Happiness data
df_2023 = pd.read_csv('../data/raw/2023.csv')  # Load the 2023 World Happiness data
df_2024 = pd.read_csv('../data/raw/2024.csv')  # Load the 2024 World Happiness data

In [42]:
# Congo variants in each dataset from 2015 to 2024

congo_2015 = df_2015[df_2015['Country'].str.contains('Congo', case=False, na=False)]
print("Congo variants in 2015 dataset:")
print(congo_2015[['Country','Happiness Rank']])

congo_2016 = df_2016[df_2016['Country'].str.contains('Congo', case=False, na=False)]
print("\nCongo variants in 2016 dataset:")
print(congo_2016[['Country','Happiness Rank']])

congo_2017 = df_2017[df_2017['Country'].str.contains('Congo', case=False, na=False)]
print("\nCongo variants in 2017 dataset:")
print(congo_2017[['Country','Happiness.Rank']])

congo_2018 = df_2018[df_2018['Country or region'].str.contains('Congo', case=False, na=False)]
print("\nCongo variants in 2018 dataset:")
print(congo_2018[['Country or region','Overall rank']])

congo_2019 = df_2019[df_2019['Country or region'].str.contains('Congo', case=False, na=False)]
print("\nCongo variants in 2019 dataset:")
print(congo_2019[['Country or region','Overall rank']])

congo_2020 = df_2020[df_2020['Country name'].str.contains('Congo', case=False, na=False)]
print("\nCongo variants in 2020 dataset:")
print(congo_2020[['Country name','Happiness Rank']])

congo_2021 = df_2021[df_2021['Country name'].str.contains('Congo', case=False, na=False)]
print("\nCongo variants in 2021 dataset:")
print(congo_2021[['Country name','Happiness Rank']])

congo_2022 = df_2022[df_2022['Country name'].str.contains('Congo', case=False, na=False)]
print("\nCongo variants in 2022 dataset:")
print(congo_2022[['Country name','Happiness Rank']])

congo_2023 = df_2023[df_2023['Country name'].str.contains('Congo', case=False, na=False)]
print("\nCongo variants in 2023 dataset:")
print(congo_2023[['Country name','Happiness Rank']])

congo_2024 = df_2024[df_2024['Country name'].str.contains('Congo', case=False, na=False)]
print("\nCongo variants in 2024 dataset:")
print(congo_2024[['Country name','Happiness Rank']])


Congo variants in 2015 dataset:
                 Country  Happiness Rank
119     Congo (Kinshasa)             120
138  Congo (Brazzaville)             139

Congo variants in 2016 dataset:
                 Country  Happiness Rank
124     Congo (Kinshasa)             125
126  Congo (Brazzaville)             127

Congo variants in 2017 dataset:
                 Country  Happiness.Rank
123  Congo (Brazzaville)             124
125     Congo (Kinshasa)             126

Congo variants in 2018 dataset:
       Country or region  Overall rank
113  Congo (Brazzaville)           114
131     Congo (Kinshasa)           132

Congo variants in 2019 dataset:
       Country or region  Overall rank
102  Congo (Brazzaville)           103
126     Congo (Kinshasa)           127

Congo variants in 2020 dataset:
            Country name  Happiness Rank
87   Congo (Brazzaville)              88
130     Congo (Kinshasa)             131

Congo variants in 2021 dataset:
           Country name  Happiness Rank
82  

**Action:**  
To resolve this, I compared the **Happiness Rank** values of both variants across the years. Since **Congo (Kinshasa)** had incomplete data in all datasets, I dropped all rows corresponding to **Congo (Kinshasa)** from the merged dataset. This ensured the dataset contains only the complete and consistent **Congo (Brazzaville)** data mapped to **`Congo`**.

In [47]:


#  Define the (Year, Happiness Rank) pairs to remove
rows_to_drop = [
    (2015, 120),
    (2016, 125),
    (2017, 126),
    (2018, 132),
    (2019, 127),
    (2020, 131),
    (2023, 133),
    (2024, 139),
]

# Build a boolean mask that is True only for those rows
mask = (
    (df['Country'] == 'Congo') &
    df.set_index(['Year', 'Happiness Rank']).index.isin(rows_to_drop)
)

#  Drop the rows
df = df[~mask]

#  Verify the result
print("Expected rows:", 131 * 10)        # 1 310
print("Actual rows after drop:", len(df))


Expected rows: 1310
Actual rows after drop: 1310


In [48]:
# Total number of unique countries
unique_countries_count = df['Country'].nunique()
print(f"Total unique countries: {unique_countries_count}")

# Number of times each country appears
country_counts = df['Country'].value_counts()

print(country_counts.to_string())


Total unique countries: 131
Country
Afghanistan                10
Romania                    10
Poland                     10
Philippines                10
Peru                       10
Paraguay                   10
Panama                     10
Palestinian Territories    10
Pakistan                   10
Norway                     10
North Macedonia            10
Nigeria                    10
Niger                      10
Nicaragua                  10
New Zealand                10
Netherlands                10
Nepal                      10
Myanmar                    10
Morocco                    10
Montenegro                 10
Mongolia                   10
Moldova                    10
Mexico                     10
Mauritius                  10
Mauritania                 10
Malta                      10
Mali                       10
Malaysia                   10
Malawi                     10
Madagascar                 10
Luxembourg                 10
Portugal                   10
Russ

In [49]:
congo_rows = df[df['Country'] == 'Congo']
print("Total 'Congo' rows:", len(congo_rows))
display(congo_rows)


Total 'Congo' rows: 10


Country  Happiness Rank  Happiness Score   Economy    Family  \
241   Congo             139            3.989  0.678660  0.662900   
243   Congo             127            4.236  0.771090  0.477990   
244   Congo             124            4.291  0.808964  0.832044   
246   Congo             114            4.559  0.682000  0.811000   
248   Congo             103            4.812  0.673000  0.799000   
250   Congo              88            5.190  0.630000  0.760000   
252   Congo              83            5.342  0.518000  0.392000   
253   Congo              99            5.075  0.950000  0.405000   
254   Congo              86            5.267  0.921000  0.665000   
256   Congo              89            5.221  0.892000  0.622000   

     Healthy life expectancy  Freedom to make life choices  \
241                 0.310510                      0.414660   
243                 0.282120                      0.379380   
244                 0.289957                      0.435026   
246                 0.343000                      0.514000   
248                 0.508000                      0.372000   
250                 0.460000                      0.390000   
252                 0.307000                      0.381000   
253                 0.355000                      0.431000   
254                 0.145000                      0.464000   
256                 0.306000                      0.523000   

     Perceptions of corruption  Generosity Continent  Year  
241                   0.116860    0.123880    Africa  2015  
243                   0.097530    0.120770    Africa  2016  
244                   0.079618    0.120852    Africa  2017  
246                   0.077000    0.091000    Africa  2018  
248                   0.093000    0.105000    Africa  2019  
250                   0.120000    0.120000    Africa  2020  
252                   0.124000    0.144000    Africa  2021  
253                   0.146000    0.130000    Africa  2022  
254                   0.136000    0.134000    Africa  2023  
256                   0.138000    0.124000    Africa  2024

Unwanted Congo (Kinshasa) rows successfully dropped.

Final dataset row count: 1,310 rows (as expected — 31 countries × 10 years).

Ensured only one valid Congo entry per year remains for analysis.

In [ ]:

#To Overwrite the Existing cleaned_df_all_2015_to_2024.csv
df.to_csv('cleaned_df_all_2015_to_2024.csv', index=False)


# Rounding Float Values to 5 Decimal Places
  To improve readability and reduce unnecessary precision, the float-type columns in the dataset have been rounded to 5 decimal places. This helps in minimizing long decimal values without significant loss of information.

In [4]:
df =pd.read_csv('../data/cleaned/cleaned_df_all_2015_to_2024.csv')

During data exploration, it was observed that the 2017 data contains float values with more than 5 decimal places, whereas the data from other years has cleaner, shorter decimal values.

To maintain consistency across all years and improve readability, the selected float columns were rounded to 5 decimal places using the pandas .round() method.

In [6]:

# Columns to round
float_cols = [
    'Happiness Score', 'Economy', 'Family',
    'Healthy life expectancy', 'Freedom to make life choices',
    'Perceptions of corruption', 'Generosity'
]

# Round those columns to 5 decimal places
df[float_cols] = df[float_cols].round(5)


In [ ]:
# to verify first 5 rows where Year is 2017
df[df['Year'] == 2017].head()


Country  Happiness Rank  Happiness Score  Economy   Family  \
2   Afghanistan             141            3.794  0.40148  0.58154   
12      Albania             109            4.644  0.99619  0.80369   
22      Algeria              53            5.872  1.09186  1.14622   
32    Argentina              24            6.599  1.18530  1.44045   
42      Armenia             121            4.376  0.90060  1.00748   

    Healthy life expectancy  Freedom to make life choices  \
2                   0.18075                       0.10618   
12                  0.73116                       0.38150   
22                  0.61758                       0.23334   
32                  0.69514                       0.49452   
42                  0.63752                       0.19830   

    Perceptions of corruption  Generosity      Continent  Year  
2                     0.06116     0.31187           Asia  2017  
12                    0.03986     0.20131         Europe  2017  
22                    0.14610     0.06944         Africa  2017  
32                    0.05974     0.10946  South America  2017  
42                    0.02667     0.08349           Asia  2017

Recalculating Happiness Rank After Data Cleaning
Some countries were removed during data cleaning, which made the original Happiness Rank inaccurate.
To address this, a new column called Happiness Rank (Cleaned Data) was created by re-ranking countries based on their Happiness Score within each year.

This ensures that the rankings reflect only the countries currently in the dataset, providing more accurate and meaningful analysis.

In [8]:
# Recalculate rank by year based on Happiness Score
df['Happiness Rank (Cleaned Data)'] = df.groupby('Year')['Happiness Score'].rank(ascending=False).astype(int)


In [9]:
df.head()


Country  Happiness Rank  Happiness Score  Economy   Family  \
0  Afghanistan             153            3.575  0.31982  0.30285   
1  Afghanistan             154            3.360  0.38227  0.11037   
2  Afghanistan             141            3.794  0.40148  0.58154   
3  Afghanistan             145            3.632  0.33200  0.53700   
4  Afghanistan             154            3.203  0.35000  0.51700   

   Healthy life expectancy  Freedom to make life choices  \
0                  0.30335                       0.23414   
1                  0.17344                       0.16430   
2                  0.18075                       0.10618   
3                  0.25500                       0.08500   
4                  0.36100                       0.00000   

   Perceptions of corruption  Generosity Continent  Year  \
0                    0.09719     0.36510      Asia  2015   
1                    0.07112     0.31268      Asia  2016   
2                    0.06116     0.31187      Asia  2017   
3                    0.03600     0.19100      Asia  2018   
4                    0.02500     0.15800      Asia  2019   

   Happiness Rank (Cleaned Data)  
0                            129  
1                            130  
2                            124  
3                            127  
4                            131

In [10]:
top_5_per_year = df[df['Happiness Rank (Cleaned Data)'] <= 5].sort_values(['Year', 'Happiness Rank (Cleaned Data)'])

# Display the top 5 countries per year
top_5_per_year[['Year', 'Country', 'Happiness Score', 'Happiness Rank (Cleaned Data)']]


Year      Country  Happiness Score  Happiness Rank (Cleaned Data)
1120  2015  Switzerland            7.587                              1
480   2015      Iceland            7.561                              2
290   2015      Denmark            7.527                              3
890   2015       Norway            7.522                              4
190   2015       Canada            7.427                              5
291   2016      Denmark            7.526                              1
1121  2016  Switzerland            7.509                              2
481   2016      Iceland            7.501                              3
891   2016       Norway            7.498                              4
361   2016      Finland            7.413                              5
892   2017       Norway            7.537                              1
292   2017      Denmark            7.522                              2
482   2017      Iceland            7.504                              3
1122  2017  Switzerland            7.494                              4
362   2017      Finland            7.469                              5
363   2018      Finland            7.632                              1
893   2018       Norway            7.594                              2
293   2018      Denmark            7.555                              3
483   2018      Iceland            7.495                              4
1123  2018  Switzerland            7.487                              5
364   2019      Finland            7.769                              1
294   2019      Denmark            7.600                              2
894   2019       Norway            7.554                              3
484   2019      Iceland            7.494                              4
834   2019  Netherlands            7.488                              5
365   2020      Finland            7.810                              1
295   2020      Denmark            7.650                              2
1125  2020  Switzerland            7.560                              3
485   2020      Iceland            7.500                              4
895   2020       Norway            7.490                              5
366   2021      Finland            7.842                              1
296   2021      Denmark            7.620                              2
1126  2021  Switzerland            7.571                              3
486   2021      Iceland            7.554                              4
836   2021  Netherlands            7.464                              5
367   2022      Finland            7.821                              1
297   2022      Denmark            7.636                              2
487   2022      Iceland            7.557                              3
1127  2022  Switzerland            7.512                              4
837   2022  Netherlands            7.415                              5
368   2023      Finland            7.804                              1
298   2023      Denmark            7.586                              2
488   2023      Iceland            7.530                              3
548   2023       Israel            7.473                              4
838   2023  Netherlands            7.403                              5
369   2024      Finland            7.741                              1
299   2024      Denmark            7.583                              2
489   2024      Iceland            7.525                              3
1119  2024       Sweden            7.344                              4
549   2024       Israel            7.341                              5

Outlier Detection and Treatment:

During data preprocessing, several numerical columns were checked for outliers using the Interquartile Range (IQR) method. This method helps detect extreme values that may skew the analysis.

IQR = Q3 − Q1, and any value outside [Q1 − 1.5×IQR, Q3 + 1.5×IQR] is considered an outlier.

To handle these outliers, we applied the Winsorization method, which limits extreme values to reduce their influence without removing any rows. Specifically, we capped the bottom and top 5% of values in each feature column.

Columns Processed:Happiness Score,Economy,Family,Healthy life expectancy,Freedom to make life choices,Perceptions of corruption and Generosity.

This approach ensures the data remains robust and stable for further statistical analysis and modeling.

In [12]:
# Columns to check for outliers
numeric_cols = [
    'Happiness Score', 'Economy', 'Family',
    'Healthy life expectancy', 'Freedom to make life choices',
    'Perceptions of corruption', 'Generosity'
]

# Dictionary to store outlier counts
outlier_summary = {}

# Loop through each column
for col in numeric_cols:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    # Count outliers
    outliers = df[(df[col] < lower) | (df[col] > upper)]
    outlier_summary[col] = {
        'Q1': round(Q1, 5),
        'Q3': round(Q3, 5),
        'Lower Bound': round(lower, 5),
        'Upper Bound': round(upper, 5),
        'Outlier Count': outliers.shape[0]
    }

# Display the results as a DataFrame
pd.DataFrame(outlier_summary).T


Q1       Q3  Lower Bound  Upper Bound  \
Happiness Score               4.72225  6.37425      2.24425      8.85225   
Economy                       0.78891  1.39413     -0.11892      2.30196   
Family                        0.87000  1.34000      0.16500      2.04500   
Healthy life expectancy       0.43510  0.77976     -0.08189      1.29674   
Freedom to make life choices  0.38000  0.59175      0.06238      0.90938   
Perceptions of corruption     0.05500  0.16394     -0.10841      0.32735   
Generosity                    0.11100  0.24600     -0.09150      0.44850   

                              Outlier Count  
Happiness Score                         2.0  
Economy                                 0.0  
Family                                 12.0  
Healthy life expectancy                 0.0  
Freedom to make life choices            8.0  
Perceptions of corruption             114.0  
Generosity                             40.0

#### Outlier Detection Using IQR Method

Outliers were detected using the Interquartile Range (IQR) method, which defines outliers as values outside the range:

**Lower Bound** = Q1 - 1.5 × IQR  
**Upper Bound** = Q3 + 1.5 × IQR  

where Q1 and Q3 are the 25th and 75th percentiles, respectively.

| Feature                        | Outlier Count | Notes                                      |
|-------------------------------|---------------|--------------------------------------------|
| Happiness Score               | 2             | Very few outliers detected                  |
| Economy                      | 0             | No outliers detected                        |
| Family                       | 12            | Some outliers present                       |
| Healthy life expectancy      | 0             | No outliers detected                        |
| Freedom to make life choices | 8             | Moderate number of outliers                 |
| Perceptions of corruption    | 114           | Many extreme outliers, high influence risk |
| Generosity                   | 40            | Significant outliers present                |

**Summary:**  
Most numerical features show minimal outliers, but *Perceptions of corruption* and *Generosity* contain many extreme values. These columns are suitable candidates for Winsorization to reduce the effect of outliers while retaining all data points.


In [ ]:
# to apply Winsorization on numerical data
from scipy.stats.mstats import winsorize

# Columns to Winsorize based on outlier counts
cols_to_winsorize = ['Perceptions of corruption', 'Generosity', 'Family', 'Freedom to make life choices']

# Create a copy of the dataframe to preserve original
df_winsorized = df.copy()

for col in cols_to_winsorize:
    # Winsorize by limiting bottom 5% and top 5% values
    df_winsorized[col] = winsorize(df_winsorized[col], limits=[0.05, 0.05])

# Check the effect by comparing original and winsorized summary stats
print("Original Perceptions of corruption stats:")
print(df['Perceptions of corruption'].describe())
print("\nWinsorized Perceptions of corruption stats:")
print(df_winsorized['Perceptions of corruption'].describe())


Original Perceptions of corruption stats:
count    1310.000000
mean        0.131393
std         0.114688
min         0.000000
25%         0.055000
50%         0.092465
75%         0.163940
max         0.587000
Name: Perceptions of corruption, dtype: float64

Winsorized Perceptions of corruption stats:
count    1310.000000
mean        0.128243
std         0.103866
min         0.016000
25%         0.055000
50%         0.092465
75%         0.163940
max         0.389000
Name: Perceptions of corruption, dtype: float64


/Users/ebenezerrajakumar/world-happiness-health-analytics/venv/lib/python3.12/site-packages/numpy/lib/function_base.py:4824: UserWarning: Warning: 'partition' will ignore the 'mask' of the MaskedArray.
  arr.partition(


In [14]:
def detect_outliers_iqr(series):
    Q1 = series.quantile(0.25)
    Q3 = series.quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = series[(series < lower_bound) | (series > upper_bound)]
    return lower_bound, upper_bound, len(outliers)

# For winsorized column
for col in cols_to_winsorize:
    lower, upper, count = detect_outliers_iqr(df_winsorized[col])
    print(f"{col} - Lower Bound: {lower:.5f}, Upper Bound: {upper:.5f}, Outliers: {count}")


Perceptions of corruption - Lower Bound: -0.10841, Upper Bound: 0.32735, Outliers: 114
Generosity - Lower Bound: -0.09150, Upper Bound: 0.44850, Outliers: 0
Family - Lower Bound: 0.16500, Upper Bound: 2.04500, Outliers: 0
Freedom to make life choices - Lower Bound: 0.06238, Upper Bound: 0.90938, Outliers: 0


/Users/ebenezerrajakumar/world-happiness-health-analytics/venv/lib/python3.12/site-packages/numpy/lib/function_base.py:4824: UserWarning: Warning: 'partition' will ignore the 'mask' of the MaskedArray.
  arr.partition(
/Users/ebenezerrajakumar/world-happiness-health-analytics/venv/lib/python3.12/site-packages/numpy/lib/function_base.py:4824: UserWarning: Warning: 'partition' will ignore the 'mask' of the MaskedArray.
  arr.partition(
/Users/ebenezerrajakumar/world-happiness-health-analytics/venv/lib/python3.12/site-packages/numpy/lib/function_base.py:4824: UserWarning: Warning: 'partition' will ignore the 'mask' of the MaskedArray.
  arr.partition(
/Users/ebenezerrajakumar/world-happiness-health-analytics/venv/lib/python3.12/site-packages/numpy/lib/function_base.py:4824: UserWarning: Warning: 'partition' will ignore the 'mask' of the MaskedArray.
  arr.partition(
/Users/ebenezerrajakumar/world-happiness-health-analytics/venv/lib/python3.12/site-packages/numpy/lib/function_base.py:4824:

In [ ]:
# Save the file after winsorization 
df_winsorized.to_csv('winsorized_df_all.csv', index=False)
